### ISSUE TITLE
-X parameter does not appear to be working

### ISSUE THREAD

**Issue by stefangordon** [ethereum-mining/ethminer#1128]
Per 
https://github.com/ethereum-mining/ethminer/issues/1116#issuecomment-390129794

I can confirm the same behavior in 14

**Comment by stefangordon** [comment]
Related to #1049 ?

**Comment by AndreaLanfranchi** [comment]
@chfast 
@jean-m-cyr 
@MariusVanDerWijden 
@ddobreff 

Mixed mode is utterly broken and frankly I can't imagine if it ever worked.
Imagine this pci configuration

| PCI Bus | Type | Vendor |
| ------ | ------- | ---- |
|01:00.0 | VGA compatible controller | AMD 
|02:00.0 | VGA compatible controller | NVIDIA 
|03:00.0 | VGA compatible controller | NVIDIA 
|04:00.0 | VGA compatible controller | AMD 
|05:00.0 | VGA compatible controller | AMD
|06:00.0 | VGA compatible controller | NVIDIA 

If a user issues ethminer -X (without specifying anything else aside the pool) he/she would expect that gpu# 1, 2, 5 to be set as CUDA miners while gpu# 0, 3, 4 as OpenCL miners.
Unfortunately this does not happen as NVIDIA cards do also support OpenCL so ethminer will firstly try to allocate OpenCL miners (all six) and then will try to allocate CUDA miners working over the already allocated OpenCL ones.

This is easily proven by ethminer -X --list-devices just ran on an NVIDIA only rig.

```
Listing OpenCL devices.
FORMAT: [platformID] [deviceID] deviceName
[1] [0] GeForce GTX 1050 Ti
	CL_DEVICE_TYPE: GPU
	CL_DEVICE_GLOBAL_MEM_SIZE: 4235919360
	CL_DEVICE_MAX_MEM_ALLOC_SIZE: 1058979840
	CL_DEVICE_MAX_WORK_GROUP_SIZE: 1024
[1] [1] GeForce GTX 1050 Ti
	CL_DEVICE_TYPE: GPU
	CL_DEVICE_GLOBAL_MEM_SIZE: 4236312576
	CL_DEVICE_MAX_MEM_ALLOC_SIZE: 1059078144
	CL_DEVICE_MAX_WORK_GROUP_SIZE: 1024
[1] [2] GeForce GTX 1050 Ti
	CL_DEVICE_TYPE: GPU
	CL_DEVICE_GLOBAL_MEM_SIZE: 4236312576
	CL_DEVICE_MAX_MEM_ALLOC_SIZE: 1059078144
	CL_DEVICE_MAX_WORK_GROUP_SIZE: 1024
[1] [3] GeForce GTX 1050 Ti
	CL_DEVICE_TYPE: GPU
	CL_DEVICE_GLOBAL_MEM_SIZE: 4236312576
	CL_DEVICE_MAX_MEM_ALLOC_SIZE: 1059078144
	CL_DEVICE_MAX_WORK_GROUP_SIZE: 1024
[1] [4] GeForce GTX 1050 Ti
	CL_DEVICE_TYPE: GPU
	CL_DEVICE_GLOBAL_MEM_SIZE: 4236312576
	CL_DEVICE_MAX_MEM_ALLOC_SIZE: 1059078144
	CL_DEVICE_MAX_WORK_GROUP_SIZE: 1024
[1] [5] GeForce GTX 1050 Ti
	CL_DEVICE_TYPE: GPU
	CL_DEVICE_GLOBAL_MEM_SIZE: 4236312576
	CL_DEVICE_MAX_MEM_ALLOC_SIZE: 1059078144
	CL_DEVICE_MAX_WORK_GROUP_SIZE: 1024

Listing CUDA devices.
FORMAT: [deviceID] deviceName
[0] GeForce GTX 1050 Ti
	Compute version: 6.1
	cudaDeviceProp::totalGlobalMem: 4235919360
	Pci: 0000:01:00
[1] GeForce GTX 1050 Ti
	Compute version: 6.1
	cudaDeviceProp::totalGlobalMem: 4236312576
	Pci: 0000:02:00
[2] GeForce GTX 1050 Ti
	Compute version: 6.1
	cudaDeviceProp::totalGlobalMem: 4236312576
	Pci: 0000:03:00
[3] GeForce GTX 1050 Ti
	Compute version: 6.1
	cudaDeviceProp::totalGlobalMem: 4236312576
	Pci: 0000:04:00
[4] GeForce GTX 1050 Ti
	Compute version: 6.1
	cudaDeviceProp::totalGlobalMem: 4236312576
	Pci: 0000:05:00
[5] GeForce GTX 1050 Ti
	Compute version: 6.1
	cudaDeviceProp::totalGlobalMem: 4236312576
	Pci: 0000:06:00
```

So one may imagine that specifying both --cuda-devices and --opencl-devices the overlapping may be mitigated. But still we get an error. In fact the index specified in any of the above mentioned cli-args is the ordinal index **whithin the subset of GPU types**.

Example. Recalling the above PCI bus sample we would write:

`ethminer -X --cuda-devices 1 2 5 --opencl-devices 0 3 4`

And this is wrong again as, for CUDA the listDevices function will return a vector sized to 3 with values {0,1,2} so there is no index 5 while on the other hand listDevices for OpenCL will return a vector for all 6 cards.

**Comment by AndreaLanfranchi** [comment]
Probably this could help

https://anteru.net/blog/2014/08/01/2483/

**Comment by chfast** [comment]
Yes, that's true. Actually that works for my single 8 GB Nvidia card :). But hashrate is not doubled :)

This comes down to cleaning up the reported devices. Reporting PCI id is possible but tricky, depends on Nvidia and AMD extensions to OpenCL.

The main idea was to always use CUDA for Nvidia cards unless opencl is forced for them with a command line argument (needed for testing). Maybe you can find a discussion about this in some issue.

**Comment by AndreaLanfranchi** [comment]
> The main idea was to always use CUDA for Nvidia cards unless opencl is forced for them with a command line argument (needed for testing). Maybe you can find a discussion about this in some issue.

I think the other way round is more feasible: OpenCL (with proper extension) may detect both NVIDIA and AMD cards while CUDA can only detect NVIDIA. So basically run a listDevices using OpenCL and subscribe GPUs with CUDA only for the ones being detected as NVIDIA.

I reckon @smurfy commented something similar in #1049

**Comment by AndreaLanfranchi** [comment]
Resolved by #1704
